### DISCUSSION TITLE
Hierachy for joints and their Rigid Bodies

### DISCUSSION THREAD

**Discussion by Conz3D** [godot-jolt/godot-jolt#692] [General]
I would like to have a basic discussion about how to structure the hierachy tree for physics systems with joints.
On the godot dicord I read all the time that you shouldn't make a rigid body a chield of another rigid body.
In my thread about my problem with the 6DOF Joint one of the first questions was about the hierarchy, too. https://github.com/godot-jolt/godot-jolt/discussions/654
I would like to have this as a separate discussion.

I never had problems with such a structure for now, when I have a joint in the hierachy with one body as parent and the 2nd body as the chield of the joint.

+ Rigid Body or Static Body (Root of the scene)
-+ Joint (with the bodies as parameter)
--+Rigid Body

I like this structure as you have a visual representation of how the bodies and the joint are connected and a Rigid Body or a Static Body is the scene root. It would be helpful with more complex systems where you have more than one Joint and more than two Rigid Bodies.

Would it be better to have a structure like this?

+ Node3D
-+ Rigid Body or Static Body
-+ Rigid Body
-+ Joint (with the bodies as parameter)

Does the hierachy influence the physics in these cases or is it not relevant for the setup of the physics?

I would prefer if the first structure would be the standard hierarchy for such a system. Maybe instead of setting the bodies in the joint as parameter, the joint is connected to the parent and the chield. Would this make sense?
With that it would be possible to attach more than one Rigid Body to another one. Like so:

+ Rigid Body or Static Body (Root of the scene)
-+ Joint (no parameter, using parent and chields)
--+Rigid Body
--+Rigid Body
--+Rigid Body

Of course only if this is possible in Jolt.

**Comment by mihe** [comment]
> On the godot dicord I read all the time that you shouldn't make a rigid body a chield of another rigid body.

As far as I know there's nothing necessarily wrong with having a physics body being the parent of another physics body. For the most part the two will act completely independent of eachother. The only real gotcha that I'm aware of is when you explicitly transform the parent in some way, in which case you will also transform the child along with it, which can sometimes be unintuitive.

EDIT: To anyone stumbling upon this in the future, there is apparently a bug in Godot at the moment where if you have a `RigidBody3D` as a child/descendant of another `RigidBody3D` it will result in a feedback loop that can cause some very strange/subtle issues. More details in [this](https://github.com/godot-jolt/godot-jolt/discussions/654#discussioncomment-7663391) comment.

> Does the hierachy influence the physics in these cases or is it not relevant for the setup of the physics?

There might be some edge case I'm not considering, so take this with a grain of salt, but to the best of my knowledge it shouldn't matter where you place a joint in the scene tree. The only thing that matters is the joint's _initial_ transform relative to the two bodies that it constrains, since that's what determines the reference frames for the two bodies as provided to the joint.

> I would prefer if the first structure would be the standard hierarchy for such a system. Maybe instead of setting the bodies in the joint as parameter, the joint is connected to the parent and the chield. Would this make sense?
> With that it would be possible to attach more than one Rigid Body to another one.

I'm not exactly sure how you see more than two bodies working, but in either case that's not a thing that Jolt supports for the types of joints that are exposed in Godot, as those are all two-body joints.

As for using child nodes instead of properties to designate the constrained bodies, there's nothing stopping you from making your own joint node, which either wraps a "real" joint node or just uses the physics server API directly, with which you can then fetch child nodes yourself and assign those as the constrained bodies.

I have no intention of changing this for the joints exposed by this extension, as I'd rather adhere to the same type of interface as the default Godot joints. I would perhaps also argue that having it be properties provides more flexibility, since you don't need to change your scene hierarchy whenever you connect/disconnect bodies from your joints.

**Reply by Conz3D** [reply]
Thank you for your answers.
Good idea, I'll try to make my own node for it, just to see how it feels.
